### PR TITLE
[onert] This draft adds unittests of nnapi_api for training

### PR DESCRIPTION
**Pull request by ragmani**
ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

**Comment by ragmani** [comment]
An error occurred on tizen build
```
[2024-06-04T04:22:04.931Z] [  827s] /home/abuild/rpmbuild/BUILD/nnfw-1.27.0/runtime/onert/backend/train/MemoryPlanner.test.cc:17:10: fatal error: gtest/gtest.h: No such file or directory
[2024-06-04T04:22:04.931Z] [  827s]    17 | #include <gtest/gtest.h>
[2024-06-04T04:22:04.931Z] [  827s]       |          ^~~~~~~~~~~~~~~
[2024-06-04T04:22:04.931Z] [  827s] compilation terminated.
[2024-06-04T04:22:04.931Z] [  827s] gmake[2]: *** [runtime/onert/backend/train/CMakeFiles/test_onert_backend_train.dir/build.make:76: runtime/onert/backend/train/CMakeFiles/test_onert_backend_train.dir/MemoryPlanner.test.cc.o] Error 1
[2024-06-04T04:22:04.931Z] [  827s] gmake[2]: *** Waiting for unfinished jobs....
[2024-06-04T04:22:04.931Z] [  827s] /home/abuild/rpmbuild/BUILD/nnfw-1.27.0/runtime/onert/backend/train/optimizer/Optimizers.test.cc:23:10: fatal error: gtest/gtest.h: No such file or directory
[2024-06-04T04:22:04.931Z] [  827s]    23 | #include <gtest/gtest.h>
[2024-06-04T04:22:04.931Z] [  827s]       |          ^~~~~~~~~~~~~~~
[2024-06-04T04:22:04.931Z] [  827s] compilation terminated.
```

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+  _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+  _context->addTrainCase(
+    uniformTCD<float>({{{1, 2, 7, 8, 3, 4, 9, 1, 5, 6, 11, 2}}}, // input dataset
+                      {{{1, -4, 1, -3, 2, -2, 2, -4}}},          // expected dataset
+                      {13.107666f}                               // last losses
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:45]: test on tensorflow with learning rate(`0.01f`)
```
Epoch 1/5
1/1 [==============================] - 0s 160ms/step - loss: 27.5685 - categorical_accuracy: 0.0000e+00
Epoch 2/5
1/1 [==============================] - 0s 2ms/step - loss: 4.6473 - categorical_accuracy: 0.5000
Epoch 3/5
1/1 [==============================] - 0s 2ms/step - loss: 3.1524 - categorical_accuracy: 0.5000
Epoch 4/5
1/1 [==============================] - 0s 2ms/step - loss: 2.3518 - categorical_accuracy: 0.5000
Epoch 5/5
1/1 [==============================] - 0s 2ms/step - loss: 1.8743 - categorical_accuracy: 0.5000
==========================
```

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+  _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+  _context->addTrainCase(
+    uniformTCD<float>({{{0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f}}}, // input dataset
+                      {{{6.5f, 7.5f, 8.5f, 3.5f, 8.5f, 5.5f, 2.5f, 3.5f}}}, // expected dataset
+                      {74.799194f}                                          // last losses
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:79]: Epoch 1/5
1/1 [==============================] - 0s 163ms/step - loss: 41.6667 - categorical_accuracy: 0.5000
Epoch 2/5
1/1 [==============================] - 0s 2ms/step - loss: 29.3928 - categorical_accuracy: 0.5000
Epoch 3/5
1/1 [==============================] - 0s 1ms/step - loss: 21.5095 - categorical_accuracy: 0.5000
Epoch 4/5
1/1 [==============================] - 0s 2ms/step - loss: 16.3874 - categorical_accuracy: 0.5000
Epoch 5/5
1/1 [==============================] - 0s 1ms/step - loss: 13.0262 - categorical_accuracy: 0.5000
==========================

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+  _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+  _context->addTrainCase(
+    uniformTCD<float>({{{0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f}}}, // input dataset
+                      {{{6.0f, 6.0f, 6.0f, 6.0f, 6.0f, 6.0f, 6.0f, 6.0f}}}, // expected dataset
+                      {72.f}                                                // last losses
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:113]:
```
Epoch 1/5
1/1 [==============================] - 0s 163ms/step - loss: 26.8223 - categorical_accuracy: 0.5000
Epoch 2/5
1/1 [==============================] - 0s 2ms/step - loss: 24.3528 - categorical_accuracy: 0.5000
Epoch 3/5
1/1 [==============================] - 0s 2ms/step - loss: 22.6278 - categorical_accuracy: 0.5000
Epoch 4/5
1/1 [==============================] - 0s 1ms/step - loss: 21.4152 - categorical_accuracy: 0.5000
Epoch 5/5
1/1 [==============================] - 0s 1ms/step - loss: 20.5577 - categorical_accuracy: 0.5000
```

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                     circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+  _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+  _context->addTrainCase(uniformTCD<float>(
+    {{{0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f}}},       // input dataset
+    {{{6.0f, 16.0f, 8.0f, 16.0f, 10.0f, 16.0f, 12.0f, 16.0f}}}, // expected dataset
+    {329.703643f}                                               // last losses
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:147]: Epoch 1/5
1/1 [==============================] - 0s 163ms/step - loss: 170.4543 - categorical_accuracy: 0.5000
Epoch 2/5
1/1 [==============================] - 0s 2ms/step - loss: 69.3072 - categorical_accuracy: 1.0000
Epoch 3/5
1/1 [==============================] - 0s 1ms/step - loss: 30.0765 - categorical_accuracy: 1.0000
Epoch 4/5
1/1 [==============================] - 0s 2ms/step - loss: 14.1054 - categorical_accuracy: 1.0000
Epoch 5/5
1/1 [==============================] - 0s 2ms/step - loss: 7.1875 - categorical_accuracy: 1.0000
==========================

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc (hunk 1 above)
Comment (jyoungyun) [reply tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:45]: TensorFlow result
```
Epoch 1/2 - loss: 6.8750
Epoch 2/2 - loss: 6.2051
```

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc (hunk 2 above)
Comment (jyoungyun) [reply tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:79]: TensorFlow result
```
Epoch 1/2 - loss: 38.0000
Epoch 2/2 - loss: 36.7402
```

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc (hunk 3 above)
Comment (jyoungyun) [reply tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:113]: TensorFlow result
(mse, sgd, learning_rate 0.001)
```
Epoch 1/2 - loss: 36.0000
Epoch 2/2 - loss: 36.0000
```

File: tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc (hunk 4 above)
Comment (jyoungyun) [reply tests/nnfw_api/src/GenModelTests/one_op_trains/DepthwiseConvolution.test.cc:147]: TensorFlow result
(mse, sgd, 0.01)
```
Epoch 1/5 - loss: 171.0000
Epoch 2/5 - loss: 69.5150
Epoch 3/5 - loss: 29.9159
Epoch 4/5 - loss: 13.7338
Epoch 5/5 - loss: 6.7380
```

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+    _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+    _context->addTrainCase(uniformTCD<float>(
+      {{{1, 3}, {0, 1, 2, 3, 4, 5, 6, 7}}, {{2, 1}, {7, 6, 5, 4, 3, 2, 1, 0}}}, // inputs
+      {{{2, 1, 5, 5, 2, 1, 5, 5}}, {{2, 1, 5, 5, 2, 1, 5, 6}}},                 // expected
+      {8.4678f}                                                                 // loss
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:50]: Epoch 1/5
2/2 [==============================] - 0s 3ms/step - loss: 9.2218 - categorical_accuracy: 0.0000e+00
Epoch 2/5
2/2 [==============================] - 0s 2ms/step - loss: 8.9554 - categorical_accuracy: 0.0000e+00
Epoch 3/5
2/2 [==============================] - 0s 1ms/step - loss: 8.7044 - categorical_accuracy: 0.0000e+00
Epoch 4/5
2/2 [==============================] - 0s 1ms/step - loss: 8.4678 - categorical_accuracy: 0.0000e+00
Epoch 5/5
2/2 [==============================] - 0s 1ms/step - loss: 8.2444 - categorical_accuracy: 0.0000e+00
==========================

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+    _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+    _context->addTrainCase(
+      uniformTCD<float>({{{0, 3}, {6, 7}}, {{5, 4}, {7, 6}}},                     // inputs
+                        {{{3, 2, 1, 2, 5, 6, 1, 0}}, {{2, 1, 5, 5, 2, 1, 5, 6}}}, // expected
+                        {12.2822f}                                                // loss
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:138]:
```
Epoch 1/5
2/2 [==============================] - 0s 3ms/step - loss: 12.5488 - categorical_accuracy: 0.0000e+00
Epoch 2/5
2/2 [==============================] - 0s 1ms/step - loss: 12.4590 - categorical_accuracy: 0.0000e+00
Epoch 3/5
2/2 [==============================] - 0s 1ms/step - loss: 12.3701 - categorical_accuracy: 0.0000e+00
Epoch 4/5
2/2 [==============================] - 0s 1ms/step - loss: 12.2822 - categorical_accuracy: 0.0000e+00
Epoch 5/5
2/2 [==============================] - 0s 1ms/step - loss: 12.1951 - categorical_accuracy: 0.0000e+00
==========================
```

**Comment by ragmani** [comment]
Done.
